### DISCUSSION TITLE
source() script at start of document

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#3250] [Q&A] [answered]
Maybe I missed an option to do this, but I'm thinking it would be nice if we could specify in the Yaml header a script to be executed at the start of the knitting process (I'm using R/knitr as a concrete example, but the same idea would presumably apply to other engines). Unlike pre-render project scripts, the purpose here is would be to set up the environment for the computations, e.g loading packages, setting options, doing some calculations for use in subsequent chunks, etc.

Of course, this is trivially achieved with a first chunk,

````
```{r setup, echo=FALSE}
library(ggplot2)
phig_ratio <- function(n){
  if(n<=2) { if(n>=0) 1 else 0 } else 1 + 1 / Recall(n-1) 
}
fig_height = 6 * phig_ratio(100)
data = read.csv('mydata.csv')
# ...
```
````

but there are some types of prose-first documents where I would prefer to have this happen less prominently, as a background process, rather than the first thing that I read when I work on the document. I find it a bit distracting, when writing prose, to be facing first not the abstract, or the first section title, introduction, etc. but obscure lines of code that are not necessarily things I should be thinking about, and which inadvertently force my brain to switch to coding mode, so to speak.

I would rather have an option in the header,

````
-----
title: ....
date: ....
init-chunk: 
  - setup.R
-----

## This is the content

In this essay, I will

```{r, fig.height = fig_height, eval = "totoro" %in% names(data) }
# ...
```
````

I ran into this when creating CVs, posters, exam papers, cover letters, etc. in Quarto; for these types of output the computations, often minimal, are there in the background to support the document and help construct it (e.g grabbing size information from image files, etc.), but I want to see them as little as possible.

**Comment by cscheid** [comment, marked as answer]
It's unlikely that we'll do this because it requires working uniformly across engines. But if your main concern is readability, you can start with an include shortcode:

```
{{< include preamble.md >}}
```

and then `preamble.md` has the `r` code cell:


````
```{r setup, echo=FALSE}
library(ggplot2)
phig_ratio <- function(n){
  if(n<=2) { if(n>=0) 1 else 0 } else 1 + 1 / Recall(n-1) 
}
fig_height = 6 * phig_ratio(100)
data = read.csv('mydata.csv')
# ...
```
````

**Reply by baptiste** [reply]
Thanks – fair enough, I can live with the workarounds (I'm using `source("setup.R")` but I should try this include, it reads more like text).

**Reply by mcanouil** [reply]
@baptiste I am going to change this to "Q&A" and mark the reply above as the answer.

**Comment by cderv** [comment]
Just adding some other solutions for **knitr** users having a R file to run as a setup chunk  (to complement other answer in this thread)

The `file` chunk option is valid and allow to use R file as code chunk content (https://bookdown.org/yihui/rmarkdown-cookbook/option-code.html)

````markdown
```{r}
#| label: setup
#| echo: false 
#| file: setup.R
```
````

This can be added in all chapter. 

Using `include` shortcode is more of a Quarto way and allows much more than a R file only as setup. 

A R only workaround for auto running some R code before any R code is using a project .Rprofile. Quarto will start a new R process for each document, and it is possible to leverage R own startup feature for this. Advanced User only I would say but quite valid anyway. To learn more: https://rstats.wtf/r-startup.html

**Reply by baptiste** [reply]
Thanks – using `.Rprofile` is a good idea; on its own I'd be worried that it's a bit too invisible (the opposite problem), but if all it does is call a more visible script at the right time with `source("_setup.R")` then it should be clear what `_setup.R` is there for.

For Python and Julia it looks less straight-forward, so perhaps a more general solution would be to extend slightly the [Environment variables machinery](https://quarto.org/docs/projects/environment.html) to have an optional STARTUP_SCRIPT that gets looked up when the computational engine starts, and if present triggers:

`R --file=$STARTUP_SCRIPT`
`julia --load $STARTUP_SCRIPT`
`python $STARTUP_SCRIPT`

**Reply by mcanouil** [reply]
Note that your `.Rprofile` can be at the root of your project next to `_quarto.yml` and be versioned using Git for example.
This is the behaviour when using `renv`.

**Reply by cderv** [reply]
I don't know for Julia and Python how that would work exactly because they are running Jupyter under the hood. Would a Python or Julia Startup script be taken into account inside Jupyter kernel execution ? 

**knitr** does execute in the R started session at the global level so this is easy, but for Python, it starts Jupyter and then start a kernel. 
Especially for Julia where we do not run Julia directly. My understanding is We start Jupyter in Python which will run a Julia Kernel. 

I don't know enough to know as of now what could be done for Python and Julia. That is also why the setup chunk or the use of `include` shortcode is the more common way of doing this. Drawback is you need to paste this `include` statement or setup chunk inside all your Qmd.

**Reply by allefeld** [reply]
An idea how to implement this in an across-engines way:

Have a YAML option `initial-block: "init.EXT"` which points to a file within the project, where EXT is the appropriate code file extension (`R`, `py`, etc.) for a language. If the file extension matches the currently running engine & kernel language, this setting is equivalent to having an executable code block

````
```{LANGUAGE}
#| include: false
{{< include init.EXT >}}
```
````

inserted before all document content.

Advantages over actually including such a block in a document:

- less boilerplate, document better readable by moving init code into a file referenced in the header
- can also be used in `_quarto.yml` or within a custom format, etc.

To avoid possibly error-prone guessing of the language from the file extension, the syntax could be

```
initial-block:
  language: r
  file: "init.R"
```

To cover several languages at once e.g. in `_quarto.yml`, this option could be nested under `knitr:` and `jupyter:` or even `kernelspec:`.

**Reply by cderv** [reply]
Thanks a lot @allefeld that is an interesting idea!

**Reply by allefeld** [reply]
Thanks!

Even simpler:

```
initial-block:
  r: "init.R"
  python: "init.py"
  …
```

The setting for a language is effective if there are any executable code blocks ` ```{language} ` already in the document.

**Comment by dchiu911** [comment]
Hi @cderv thanks for the suggestions.

Maybe I should be asking this in rstudio issue but do you think it would make sense for RStudio to allow "Run Selected Chunk" to actually evaluate the following chunk and run `setup.R` when the file option is presented

`````
```{r}
#| label: setup
#| file: setup.R
```
`````
or we might as well just use `source()` for that purpose? 
`````
```{r}
#| label: setup
source("setup.R")
```
`````
I tried using the include shortcode to adhere to quarto syntax but evaluation also does not occur 

```
{{< include _setup.qmd >}}
```

**Reply by cderv** [reply]
> do you think it would make sense for RStudio to allow "Run Selected Chunk" to actually evaluate the following chunk and run setup.R when the file option is presented

Yes `file` option is a valid knitr option, and that is equivalent to fill the cell with content of the file to then evaluate as a normal cell. 

When `echo: true`, the source code cell in the output will have the content of the file shown. 

> we might as well just use source() for that purpose?

This is a bit different in the sense that **knitr** does not really do `source()`, and the code cell in the output will be `source("setup.R")`. 

But the purpose is the same indeed and using source is fine. 

To be the safest, you should follow advice here and set the `local` argument
https://bookdown.org/yihui/rmarkdown-cookbook/source-script.html

> I tried using the include shortcode to adhere to quarto syntax but evaluation also does not occur

It should if your `_setup.qmd` content contains a R chunk with your setup content. So any of the two solutions you suggested would work. 

What is the content that does not work ?

**Reply by dchiu911** [reply]
Maybe I should be more explicit:
```
{{< include _A.qmd >}}

{{< include _B.qmd >}}

{{< include _C.qmd >}}
```
are evaluated when I `quarto render`. But when I "Run All" in their parent `qmd` interactively, they are not evaluated. Is this expected behaviour? Is there a solution where it would work both in command line and interactively? That's why I thought using `source()` would faciliate this.

**Reply by mcanouil** [reply]
That's the job of the IDE, but this is not that easy because it means for the IDE to execute the include shortcode somehow and without actually rendering the document.

**Reply by cderv** [reply]
> But when I "Run All" in their parent qmd interactively, they are not evaluated. Is this expected behaviour?

@dchiu911 Report this feature request to the RStudio IDE, or any other IDE you support. If this is VSCODE with Quarto extension, then feature request should be at https://github.com/quarto-dev/quarto

As Mickaël said - not that easy. 

> That's why I thought using source() would faciliate this.

Regarding this interactive constraint, yes using `source()` in a cell is probably the best. I don't think RStudio IDE does support `file:` cell option yet. 

All interactive feature you have in IDEs are not part of Quarto CLI - they are third party support to integrate Quarto and IDE features together.
